Add role to deploy 17.1 env for adoption

cescgina opened the pull request:
Add a new role that will deploy a tripleo environment that will serve as
source for adoption. This role is expected to cosume the infra created
by [1], and a 17.1 scenario definition from the data-plane-adoption
repo, introduced by [2].

It also introduce a small fix to the deploy-ocp.yml so the resulting ocp
cluster is ready (the nodes needed to be uncordoned).

[1] https://github.com/openstack-k8s-operators/ci-framework/pull/2285
[2] https://github.com/openstack-k8s-operators/data-plane-adoption/pull/597

cjeanner commented on roles/adoption_osp_deploy/files/undercloud.conf:
I'd rather consume the one provided by the package, and use `community.general.ini_file` to edit the entries.

cescgina replied:
which package provides the undercloud.conf? I added it to patch it with the ini_file module as you suggested, that seemed like a good idea

cjeanner commented on roles/adoption_osp_deploy/tasks/deploy_ceph.yml:
```diff
@@ -0,0 +1,71 @@
+# License for the specific language governing permissions and limitations
+# under the License.
+
+- name: Create block devices on compute nodes using ceph playbook
```
in our case, we can even go with actual drives (via the extra_disks parameter in the VMs description). Would be faster.

cescgina replied:
I started by porting what we currently have, but yes, we should align it better to what we currently have in the VA

cescgina replied:
the last version uses extra_disks instead of the block devices, it seems to work well

cjeanner commented on roles/adoption_osp_deploy/tasks/deploy_ceph.yml:
I'd rather probably rely on a hook to deploy ceph. It's not needed in all cases, and could be a pre_oc_run task for instance, or something like that.

cescgina replied:
hooks are good idea, I'll try to introduce them for the next update

bogdando replied:
+1 pre_oc_run, we will also need it in multi-stack scenarios

cjeanner commented on roles/adoption_osp_deploy/tasks/overcloud_deploy.yml:
```diff
@@ -0,0 +1,58 @@
+         _cifmw_scenario_content.adoption.overcloud_services_file
+        ] | path_join
+      }}
+    _overcloud_deploy_cmd: >-
```
I'd be really more flexible here about what we can pass. Reduce the default parameters to a bare minimum, and let others pass whatever they need from the automation content.

cescgina replied:
sounds good, I'll move more of them to the architecture vars files

cjeanner commented on roles/adoption_osp_deploy/tasks/main.yml:
```diff
@@ -0,0 +1,34 @@
+# under the License.
+
+- name: Prepare enviornment for OSP deployment
+  ansible.builtin.import_tasks: prepare.yml
```
I think this file is missing?

cescgina replied:
yep, haven't gotten to this bit yet

cjeanner commented on roles/adoption_osp_deploy/tasks/main.yml:
```diff
@@ -0,0 +1,34 @@
+  ansible.builtin.import_tasks: prepare.yml
+
+- name: Generate configuration files
+  ansible.builtin.import_tasks: config_files.yml
```
ditto - missing

cjeanner reviewed (commented):
Would need some more flexibility, but it could match the need with some tweaks/iterations.

ciecierski commented on deploy-osp-adoption.yml:
```diff
@@ -0,0 +1,93 @@
+
+    - name: Read inventory groups
+      ansible.builtin.slurp:
+        path: "/home/zuul/ci-framework-data/reproducer-inventory/all-group.yml"
```
cifmw_basedir var should be used instead of this hard-coded value

cescgina replied:
Done, thanks

ciecierski commented on deploy-osp-adoption.yml:
```diff
@@ -0,0 +1,93 @@
+        - name: Read groups inventory
+          vars:
+            _group_name: "{{ item.key[:-1] }}"
+            _path_inventories: "/home/zuul/ci-framework-data/reproducer-inventory"
```
the same as comment above, use {{ cifmw_basedir }}

cescgina replied:
Done

ciecierski commented on roles/adoption_osp_deploy/tasks/prepare_undercloud.yml:
```diff
@@ -0,0 +1,242 @@
+        {%- set _ = vms.update({vm: 'controllers'}) -%}
+      {%- endfor -%}
+      {{ vms }}
+  ansible.builtin.include_role:
```
ansible_libvirt_pools fact expected by in start_vms.yml is not defined at this point and must be defined. 
I see there is task "Gather pool fact" in libvirt_manage role which builds this variable. Maybe it there should be the same task before starting tasks from start_vms.yml

cescgina replied:
added the var, thanks

ciecierski commented on roles/adoption_osp_deploy/tasks/prepare_overcloud.yml:
```diff
@@ -0,0 +1,171 @@
+          - fast-datapath-for-rhel-9-x86_64-rpms
+          - rhceph-6-tools-for-rhel-9-x86_64-rpms
+      become: true
+      ansible.builtin.command:
```
Before running this command nodes must be registered RHSM. For manual run of the playbook we nodes could be registered using redhat_subscription module[1]
[1]https://docs.ansible.com/ansible/latest/collections/community/general/redhat_subscription_module.html

cescgina replied:
I've added now a task to register nodes when credentials are passed

cjeanner commented on deploy-ocp.yml:
Maybe it would deserves its own PR, since it's not directly related to the 17.1 env deploy

cjeanner commented on roles/adoption_osp_deploy/README.md:
```diff
@@ -0,0 +1,34 @@
+deployment. Defaults to `pool.ntp.org`
+* `cifmw_adoption_osp_deploy_repos`: (List) List of 17.1 repos to enable. Defaults to
+`[rhel-9-for-x86_64-baseos-eus-rpms, rhel-9-for-x86_64-appstream-eus-rpms, rhel-9-for-x86_64-highavailability-eus-rpms, openstack-17.1-for-rhel-9-x86_64-rpms, fast-datapath-for-rhel-9-x86_64-rpms, rhceph-6-tools-for-rhel-9-x86_64-rpms]`
+* `cifmw_adoption_osp_deploy_skip_stages`: (String or List) Stages to skip
```
hmm I'd use actual `tags: ['undercloud', 'overcloud']` and leverage `--skip-tags`. That would be better, since it would really skip the tasks without even having to consider the actual tasks with some `when` condition.

UNLESS this would be needed in CI context? Not sure if Zuul job definition allows to pass down and tags to skip... ?

cescgina replied:
I can't think of any use case for this in a CI job, I added it mostly to avoid having to re-run the undercloud deployment whenever something went wrong with the overcloud deploy. I remembered this was implemented in kustomize_deploy and copied the idea from there. But you're right that this could be much simpler with tags, I'll give it a try locally and update the PR

cescgina replied:
I've updated the PR moving to ansible tags, it's indeed simpler, thanks @cjeanner

cjeanner commented on roles/adoption_osp_deploy/tasks/main.yml:
```diff
@@ -0,0 +1,155 @@
+  ansible.builtin.import_tasks: config_files.yml
+
+- name: Group undercloud tasks
+  vars:
```
so if you're able to go with `tags: undercloud` instead, that would make things easier.

cescgina replied:
done

cjeanner commented on roles/adoption_osp_deploy/tasks/main.yml:
```diff
@@ -0,0 +1,155 @@
+        msg: "Failing on demand {{ cifmw_adoption_osp_deploy_stopper }}"
+
+- name: Group overcloud tasks
+  vars:
```
ditto, `tags: undercloud` would look nicer imho - but beware of zuul context as put in the first comment.

marios commented on scenarios/adoption/hci.yml:
```diff
@@ -55,14 +55,27 @@ libvirt_manager_patch_layout:
+      mtu: 1496
   group-templates:
-    osp-controllers:
+    computes:
```
which computes is this then in the adoption jobs. I mean, osp_computes is the source computes which then become the destination data plane nodes.

cescgina replied:
these are the computes from the greenfield jobs, we don't use them for adoption (in https://github.com/openstack-k8s-operators/ci-framework/pull/2297/files#diff-60530f309d776a3851cf8d6bfb03bf6c4bb531cc5de3f57e66f5a14021d545f2R27 we set the number of vms of this group to 0). But, since we reuse the base network definition https://github.com/openstack-k8s-operators/ci-framework/blob/main/scenarios/reproducers/networking-definition.yml  we can't undefine anything that is already there. So I need to make sure the ips for unused compute group do not overlap with the ones from osp-compute groups. I'll add a comment clarifying this

marios commented on roles/adoption_osp_deploy/tasks/login_registries.yml:
```diff
@@ -0,0 +1,50 @@
+  no_log: true
+  ansible.builtin.command: >-
+    subscription-manager register --force
+    --org "{{ cifmw_adoption_osp_deploy_rhsm_org }}"
```
how will this work though i mean the secret cannot live in this repo as it is not trusted/config repo.

cescgina replied:
this is not intended to be used in CI jobs, this is here in case someone wants to run the deployment manually like I've been doing to test it. Once I start building the zuul jobs, I'll need to do something similar from a config repo

marios commented on roles/adoption_osp_deploy/tasks/login_registries.yml:
```diff
@@ -0,0 +1,50 @@
+      ansible.builtin.command: >
+        podman login
+        --username "{{ cifmw_adoption_osp_deploy_container_user }}"
+        --password "{{ cifmw_adoption_osp_deploy_container_password }}"
```
same this will have to be moved to where these secrets live

marios commented on deploy-ocp.yml:
```diff
@@ -67,3 +67,21 @@
+        cifmw_openshift_kubeconfig: >-
+          {{ (_auth_path, 'kubeconfig') | ansible.builtin.path_join }}
+      ansible.builtin.include_role:
+        name: openshift_adm
```
this one was moved into a new review/needs rebase https://github.com/openstack-k8s-operators/ci-framework/pull/2413

cescgina replied:
yep, will rebase once 2413 merges

marios reviewed (commented):
looks great, biggest issue i saw was the secrets that needs to be worked out

fmount commented on roles/adoption_osp_deploy/templates/tripleo-ansible-inventory.yaml.j2:
```diff
@@ -0,0 +1,36 @@
+    {% endfor %}
+computes:
+  children:
+    {{ _adoption_source_scenario.roles_groups_map['osp-computes'] }}: {}
```
I'm wondering if this inventory can be used in place of [1] when we trigger the ceph migration (after the adoption) and a playbook that is meant to prepare the TripleO environment [2].
This would save us time to build a "ceph compatibility" layer when it's about executing those playbooks. Do you think we can extend this to have a Ceph section? Maybe worth discussing this in a follow up.

[1] https://review.rdoproject.org/r/c/rdo-jobs/+/53695/59/playbooks/data_plane_adoption/templates/ceph_inventory.j2
[2] https://github.com/openstack-k8s-operators/data-plane-adoption/pull/637

cescgina replied:
@fmount I can try adding a new section matching what you have in `ceph_inventory` once I finish the current testing run, and check that it doesn't interfere with what I currently have. If that works we can get it in on this first version, if we need further changes we can leave it for a follow-up

fmount replied:
ack and thank you!

cescgina replied:
@fmount I think I was able to make the change, and the overcloud deployed fine, here is the resulting inventory, let me know if that looks ok for your usecase:

```
ComputeHCI:
  hosts:
    osp-compute-l6eicnz2-0:
      ansible_host: 192.168.122.106
      canonical_hostname: osp-compute-l6eicnz2-0.localdomain
      ctlplane_ip: 192.168.122.106
      external_ip: 172.21.0.106
      internal_api_ip: 172.17.0.106
      storage_ip: 172.18.0.106
      storagemgmt_ip: 172.20.0.106
      tenant_ip: 172.19.0.106
    osp-compute-l6eicnz2-1:
      ansible_host: 192.168.122.107
      canonical_hostname: osp-compute-l6eicnz2-1.localdomain
      ctlplane_ip: 192.168.122.107
      external_ip: 172.21.0.107
      internal_api_ip: 172.17.0.107
      storage_ip: 172.18.0.107
      storagemgmt_ip: 172.20.0.107
      tenant_ip: 172.19.0.107
    osp-compute-l6eicnz2-2:
      ansible_host: 192.168.122.108
      canonical_hostname: osp-compute-l6eicnz2-2.localdomain
      ctlplane_ip: 192.168.122.108
      external_ip: 172.21.0.108
      internal_api_ip: 172.17.0.108
      storage_ip: 172.18.0.108
      storagemgmt_ip: 172.20.0.108
      tenant_ip: 172.19.0.108
  vars:
    ansible_ssh_common_args: -o StrictHostKeyChecking=no
    ansible_ssh_user: zuul
Controller:
  hosts:
    osp-controller-l6eicnz2-0:
      ansible_host: 192.168.122.103
      canonical_hostname: osp-controller-l6eicnz2-0.localdomain
      ctlplane_ip: 192.168.122.103
      external_ip: 172.21.0.103
      internal_api_ip: 172.17.0.103
      storage_ip: 172.18.0.103
      storagemgmt_ip: 172.20.0.103
      tenant_ip: 172.19.0.103
    osp-controller-l6eicnz2-1:
      ansible_host: 192.168.122.104
      canonical_hostname: osp-controller-l6eicnz2-1.localdomain
      ctlplane_ip: 192.168.122.104
      external_ip: 172.21.0.104
      internal_api_ip: 172.17.0.104
      storage_ip: 172.18.0.104
      storagemgmt_ip: 172.20.0.104
      tenant_ip: 172.19.0.104
    osp-controller-l6eicnz2-2:
      ansible_host: 192.168.122.105
      canonical_hostname: osp-controller-l6eicnz2-2.localdomain
      ctlplane_ip: 192.168.122.105
      external_ip: 172.21.0.105
      internal_api_ip: 172.17.0.105
      storage_ip: 172.18.0.105
      storagemgmt_ip: 172.20.0.105
      tenant_ip: 172.19.0.105
  vars:
    ansible_ssh_common_args: -o StrictHostKeyChecking=no
    ansible_ssh_user: zuul
Undercloud:
  hosts:
    undercloud: {}
  vars:
    ansible_connection: local
    ansible_host: localhost
allovercloud:
  children:
    ComputeHCI: {}
    Controller: {}
computes:
  children:
    ComputeHCI: {}
overcloud:
  hosts:
    osp-controller-l6eicnz2-0:
      ansible_user: tripleo-admin
      ansible_host: 192.168.122.103
    osp-controller-l6eicnz2-1:
      ansible_user: tripleo-admin
      ansible_host: 192.168.122.104
    osp-controller-l6eicnz2-2:
      ansible_user: tripleo-admin
      ansible_host: 192.168.122.105
    osp-compute-l6eicnz2-0:
      ansible_user: tripleo-admin
      ansible_host: 192.168.122.106
    osp-compute-l6eicnz2-1:
      ansible_user: tripleo-admin
      ansible_host: 192.168.122.107
    osp-compute-l6eicnz2-2:
      ansible_user: tripleo-admin
      ansible_host: 192.168.122.108
```

fmount replied:
awesome thank you! By doing this after the overcloud is deployed we can still point to the ceph bits w/o major changes!

fmount commented on roles/adoption_osp_deploy/templates/os_net_config_overcloud.yml.j2:
```diff
@@ -0,0 +1,42 @@
+  - {{ gateway_ip }}
+  domain: []
+  addresses:
+  - ip_netmask: {{ ctlplane_ip }}/{{ ctlplane_cidr }}
```
I recently had intermittent ssh issues with this config, while having the ctlplane ip (that I assume we use to ssh to the node) on a dedicated nic didn't present such problem. 
I'm not asking to change this layout at this point, but just wanted to point out an issue that I faced "a lot" during [1] testing, while not present in the infrared based jobs where the ssh interface is on a dedicated nic.

[1] https://review.rdoproject.org/r/c/testproject/+/53696

cescgina replied:
@fmount tbh at first I thought this would pose the problems you mentioned, but it seems to work fine in the machine I'm currently testing on. How does the config look like when you have the controlplane ip in a dedicated nic? What ip do you assign to the bridge?

fmount replied:
I have to check how the downstream env looks like, I just faced that issue during the ceph migration playbook, when you need to run `os-net-config` to update the nodes' net config.
The interesting part is that I'm not facing it anymore in my last attempts, so I'm confused about this one.
I think downstream we simply do not have a `ovs_bridge` for the ssh net nic.
Example:

```
1: lo    inet 127.0.0.1/8 scope host lo\       valid_lft forever preferred_lft forever
2: enp1s0    inet 192.168.24.8/24 brd 192.168.24.255 scope global enp1s0\       valid_lft forever preferred_lft forever
2: enp1s0    inet 192.168.24.54/32 brd 192.168.24.255 scope global enp1s0\       valid_lft forever preferred_lft forever
7: br-ex    inet 10.0.0.119/24 brd 10.0.0.255 scope global br-ex\       valid_lft forever preferred_lft forever
8: vlan30    inet 172.17.3.109/24 brd 172.17.3.255 scope global vlan30\       valid_lft forever preferred_lft forever
8: vlan30    inet 172.17.3.135/32 brd 172.17.3.255 scope global vlan30\       valid_lft forever preferred_lft forever
9: vlan70    inet 172.17.5.142/24 brd 172.17.5.255 scope global vlan70\       valid_lft forever preferred_lft forever
9: vlan70    inet 172.17.5.63/32 brd 172.17.5.255 scope global vlan70\       valid_lft forever preferred_lft forever
10: vlan50    inet 172.17.2.69/24 brd 172.17.2.255 scope global vlan50\       valid_lft forever preferred_lft forever
11: vlan20    inet 172.17.1.119/24 brd 172.17.1.255 scope global vlan20\       valid_lft forever preferred_lft forever
12: vlan40    inet 172.17.4.131/24 brd 172.17.4.255 scope global vlan40\       valid_lft forever preferred_lft forever
```
and

```
---
network_config:
- type: interface
  name: nic1
  use_dhcp: false
  dns_servers: ['172.16.0.1', '10.0.0.1']
  domain: []
  addresses:
  - ip_netmask: 192.168.24.8/24
  routes: [{'default': True, 'nexthop': '192.168.24.1'}]
- type: vlan
  vlan_id: 20
  device: nic1
  addresses:
  - ip_netmask: 172.17.1.119/24
  routes: []
- type: vlan
  vlan_id: 40
  device: nic1
  addresses:
  - ip_netmask: 172.17.4.131/24
  routes: []
- type: ovs_bridge
  name: br-tenant
  use_dhcp: false
  members:
  - type: interface
    name: nic2
    primary: true
  - type: vlan
    vlan_id: 50
    addresses:
    - ip_netmask: 172.17.2.69/24
    routes: []
  - type: vlan
    vlan_id: 30
    addresses:
    - ip_netmask: 172.17.3.109/24
    routes: []
  - type: vlan
    vlan_id: 70
    addresses:
    - ip_netmask: 172.17.5.142/24
    routes: []
- type: ovs_bridge
  name: br-ex
  dns_servers: ['172.16.0.1', '10.0.0.1']
  domain: []
  use_dhcp: false
  addresses:
  - ip_netmask: 10.0.0.119/24
  routes: []
  members:
  - type: interface
    name: nic3
    primary: true
```

Not sure it helps and you don't have to change what you have because of this, but hope this gives you the idea.

cescgina replied:
thanks that's helpful. It's quite different to what I have and tbh I'm not sure how much would I need to change on the tripleo side to use something like this, for now I'm leaning on keeping the current one and having this discussion here for reference in case we start seeing problems with the interface. Now I wonder though, is there some topology where the current setup will be problematic and will need to be changed? Should we have this template as part of the scenario definition?

fmount replied:
I'm not sure, maybe there's not, so I think we can continue with the current implementation and double check or provide overrides later if necessary

fmount reviewed (commented):
Thank you @cescgina for starting this. I think we can have a testproject to prove parity with what we have in rdo and then start merging your changes. They might be eventually refined w/ follow up patches if required.

marios reviewed (commented):
as discussed last week we should try and get it merged this week and then iterate on further fixes and improvements

lewisdenny commented on hooks/playbooks/adoption_deploy_ceph.yml:
```diff
@@ -0,0 +1,138 @@
+          {{
+            _adoption_source_scenario.stacks[stack_index | int]
+          }}
+        _overcloud_name: >-
```
suggestion: I see the overcloud name is coming from [1], you are setting a default here and in other places for `stack.stackname` but it seems all the other vars coming from the scenario are left in this role without default. 

I think it would be best to have the vars defined in one place, the scenario. And I think it would be good to have an assert in the adoption_osp_deploy role that everything mandatory is provided.  

[1] https://github.com/openstack-k8s-operators/data-plane-adoption/blob/43713883a05bb2fdfecb3fb19f24fd594051fc41/scenarios/hci.yaml#L36

cescgina replied:
indeed, this default is a leftover from the beginning that I forgot about, I'll remove it and try to add some validation.

lewisdenny commented on roles/adoption_osp_deploy/tasks/main.yml:
```diff
@@ -0,0 +1,140 @@
+- name: Deploy undercloud
+  tags:
+    - undercloud
+  ansible.builtin.import_tasks: deploy_undercloud.yml
```
Question: I note this task file has a single task in it, are you planning to call it from other places? Is there a reason you extracted it to it's own file?

cescgina replied:
No, I don't expect to call it anywhere else. When I started writing the role it had more tasks, but then I refactored some of them out and it was left with one task. I'll move the task to the main file, since it's quite small

lewisdenny reviewed (changes requested):
Hi @cescgina 

This is a massive role mate, please forgive me if I have missed anything. I can tell this would have been a tonne of work... thank you for your massive effort!

cescgina commented on roles/adoption_osp_deploy/README.md:
```diff
@@ -0,0 +1,32 @@
+Deploy OSP 17.1 environment for adoption based on DTs.
+
+## Privilege escalation
+None
```
good catch, I completely forgot about this part, we do actually need that, I'll update the README

cescgina commented on roles/adoption_osp_deploy/tasks/prepare_undercloud.yml:
```diff
@@ -0,0 +1,217 @@
+- name: Generate os-net-config file
+  become: true
+  delegate_to: "osp-undercloud-0"
+  vars:
```
indeed, they are all internal, I'll update them all to have an underscore

cescgina commented on deploy-osp-adoption.yml:
```diff
@@ -0,0 +1,108 @@
+# "create-infra.yml"
+
+- name: Deploy 17.1 for adoption
+  hosts: 'localhost'
```
this one is a bit tricky, I need to use the `cifmw_target_host` var because some tasks need to be delegated to the hypervisor (like starting vms or fetching the network env defintion) but the playbook itself needs to run from localhost otherwise it does not find the hook to deploy ceph (I don't have ci-framework on the hypervisor)

lewisdenny replied:
ack, understood mate.

cescgina commented on deploy-osp-adoption.yml:
```diff
@@ -0,0 +1,108 @@
+        name: _adoption_source_scenario
+
+    - name: Gather ansible_user_id from {{ cifmw_target_host }}
+      delegate_to: "{{ cifmw_target_host }}"
```
good catch, will add it

cescgina commented on roles/adoption_osp_deploy/README.md:
```diff
@@ -0,0 +1,32 @@
+- Before deploying overcloud: `before_deploy_overcloud`
+- After deploying overcloud: `after_deploy_overcloud`
+
+## Examples
```
I'll add it in the next update

cescgina commented on roles/adoption_osp_deploy/tasks/login_registries.yml:
```diff
@@ -0,0 +1,50 @@
+# License for the specific language governing permissions and limitations
+# under the License.
+
+- name: Activate red hat subscription
```
done, thanks!

cescgina commented on roles/adoption_osp_deploy/tasks/login_registries.yml:
```diff
@@ -0,0 +1,50 @@
+    - cifmw_adoption_osp_deploy_container_password is defined
+    - cifmw_adoption_osp_deploy_container_registry is defined
+  block:
+    - name: Install podman for container registry login
```
good point, I can use the `podman` role for that, will make the change

cescgina replied:
@lewisdenny I tried using the `podman` role, but that uses the `package` module instead of `dnf` and I'm hitting this issue https://github.com/ansible/ansible/issues/82598 I'm not sure what is the problem, but tbh I don't think it's worth it to try and find out the problem

lewisdenny replied:
No you are right, not worth it!

cescgina commented on roles/adoption_osp_deploy/tasks/prepare_overcloud.yml:
```diff
@@ -0,0 +1,236 @@
+    - name: Generate os-net-config file for overcloud nodes
+      become: true
+      delegate_to: "{{ overcloud_vm }}"
+      vars:
```
will update and make them all internal

ciecierski commented on deploy-osp-adoption.yml:
```diff
@@ -0,0 +1,108 @@
+    - name: Set inventory_file for localhost to use with hooks
+      ansible.builtin.set_fact:
+        inventory_file: "{{ hostvars[cifmw_target_host]['inventory_file'] }}"
+
```
If cifmw_target_host is not mandatory, then this shoul be 
cifmw_target_host | default('localhost')

cescgina replied:
thanks @ciecierski I pushed an update that fixes this issue

cescgina commented:
@lewisdenny thanks for the extensive review, I've updated the PR with some changes, please take a look when you have some time

pablintino commented on deploy-osp-adoption.yml:
```diff
@@ -0,0 +1,115 @@
+            _path_inventories: "{{ ansible_user_dir }}/ci-framework-data/reproducer-inventory"
+          ansible.builtin.slurp:
+            path: "{{ _path_inventories }}/{{ _group_name }}-group.yml"
+          register: _files
```
Meaningless variable name

cescgina replied:
thanks, I made the name more explicit

pablintino commented on deploy-osp-adoption.yml:
```diff
@@ -0,0 +1,115 @@
+          vars:
+            _content: "{{ _inventory_file.content | b64decode | from_yaml }}"
+            _group_name: "{{ _inventory_file.source | basename | regex_replace('-group.yml', '') }}"
+            _nodes: "{{ _content[_group_name~'s']['hosts'].keys() }}"
```
Assuming the groups will be always named _group_name + 's' looks fragile. One small change in the reproducer inventory dump and we will miss that adoption is making this assumption.

cescgina replied:
indeed, that is error-prone and turns out the assumption is not needed, I've changed the code to simply read the group name from the file

pablintino commented on hooks/playbooks/adoption_deploy_ceph.yml:
```diff
@@ -0,0 +1,143 @@
+      block:
+        - name: Copy ceph osd file
+          delegate_to: "osp-undercloud-0"
+          ansible.builtin.copy:
```
File manipulations without explicitly setting the `mode` will cause linter issues. iirc it's one of the rules we disabled, so no problem for now, but at some point someone will need to touch this freshly added code to make it complaint.

cescgina replied:
good catch, I haven't thought about this, I think I added the `mode` to all the tasks now

pablintino commented on roles/adoption_osp_deploy/tasks/config_files.yml:
```diff
@@ -0,0 +1,176 @@
+
+- name: Generate config download file
+  vars:
+    _config_download: >
+      {{
+        _original_config_download['content'] |
+        b64decode | from_yaml
+      }}
+    _new_config_download_fields: {}
+    _ctlplane_net: "{{ cifmw_networking_env_definition.networks.ctlplane }}"
```
All the `*download*` vars are not used till the very end of the block in one single task. Care to move them to there? I was searching where are they used all over the file...

cescgina replied:
yep, moved them all to the end

pablintino commented on roles/adoption_osp_deploy/tasks/config_files.yml:
```diff
@@ -0,0 +1,176 @@
+        ansible.utils.keep_keys(target=_stack.stack_nodes)
+      }}
+  block:
+    - name: Gather overcloud tripleo nodes
+      when: group.key is in _hostname_map_translation
+      ansible.builtin.set_fact:
+        _tripleo_nodes: >-
+          {{
+             _tripleo_nodes | default([]) +
+             group.value
+          }}
+      loop: "{{ _vm_groups | dict2items }}"
+      loop_control:
+        loop_var: group
+        label: "{{ group.key }}"
+
+    - name: Generate Hostnamemap field
+      when: group.key is in _hostname_map_translation
+      vars:
+        _tripleo_name: "{{ _hostname_map_translation[group.key] }}"
+        _group_nodes: >-
+          {%- set hosts = {} -%}
+          {%- for node in group.value -%}
+            {%- set key = _tripleo_name ~ '-' ~ loop.index0 | string  -%}
+            {%- set _ = hosts.update({key: node}) -%}
+          {%- endfor -%}
+          {{ hosts }}
+      ansible.builtin.set_fact:
+        _hostname_map: >-
+          {{
+            _hostname_map | default({}) |
+            combine(_group_nodes)
+          }}
+      loop: "{{ _vm_groups | dict2items }}"
+      loop_control:
+        loop_var: group
+        label: "{{ group.key }}"
```
Both loops can be perfectly joined and the second one doesn't depend in the computation done in the first one.

cescgina replied:
indeed, thanks, I've merged them

pablintino commented on roles/adoption_osp_deploy/tasks/prepare_undercloud.yml:
```diff
@@ -0,0 +1,217 @@
+  args:
+    apply:
+      delegate_to: "osp-undercloud-0"
+
```
Almost all the following tasks are delegated. Why not use a block to improve readability?

cescgina replied:
done, thanks!

pablintino reviewed (commented):
I did my best to review this one (not finished yet), but with this size is almost impossible to do a proper review.
Most of my comments are cosmetic ones, but I'm writing down them mainly because I'm sure that no one will have time to fix them when this PR is merged.

jistr commented:
There are already detailed reviews so i just did a higher level "let me grasp what's going on" review.

I just wondered whether we'll want to support deployment without `subscription-manager` (with `rhos-release`), but that's yet another potential enhancement that should probably **not** be included in this PR anymore :)

Thank you very much for the work on this. LGTM :+1:

lewisdenny commented:
I will take another pass after @pablintino review is addressed

cescgina commented on roles/adoption_osp_deploy/tasks/prepare_overcloud.yml:
```diff
@@ -0,0 +1,243 @@
+         _stack.network_data_file
+        ] | path_join
+      }}
+    _network_data_file_dest: "{{ ansible_user_dir }}/network_data_{{ _overcloud_name }}.yaml"
```
done, thanks

cescgina commented on deploy-osp-adoption.yml:
```diff
@@ -0,0 +1,115 @@
+          cifmw_adoption_osp_deploy_scenario:
+            {{ cifmw_adoption_osp_deploy_scenario | to_nice_yaml(indent=2) | indent(width=2) }}
+          _vm_groups:
+            {{ _vm_groups | to_nice_yaml(indent=2) | indent(width=2) }}
```
thanks, that is a great improvement!

cescgina commented on deploy-osp-adoption.yml:
```diff
@@ -0,0 +1,115 @@
+            loop_var: "_inventory_file"
+
+    - name: Gather ansible_user_id from controller for hooks
+      ansible.builtin.setup:
```
fixed the task name, it was doing the right thing, but the description was wrong

bogdando replied:
I am late to the party, albeit my question is likely related to the ci framework design, not this implementation that follows it. Why do we need to make ansible parsing other ansibler objects from yaml files, instead of using direct inputs?

cescgina commented:
@pablintino thanks for the review, I think I've addressed all your comments @lewisdenny I've updated the PR with the changes after testing locally, ptal when you have some time

pablintino commented:
/approve

frenzyfriday commented:
/lgtm
